# Part 2: Car Factors 

You are to construct a predictive model that provides the duration till sold for a given model.  

1. Populate the carsfactors.py following the hints in the comments
1. Integrate with carfactors_service.py
1. Test locally
1. Build requirements.txt and Dockerfile
1. Build a docker image
1. Test Locally
1. Push to docker hub
1. Populate readme for both github and docker hub (with example docker commands)
1. Populate this notebook with working output and a summary that contains an impression of the model and how to improve it.

* ***Review the [codeSamplesforCategoricalData.ipynb](./codeSamplesforCategoricalData.ipynb) for code review of the categorical data manipulations***.

In [ ]:
# import sys
# !{sys.executable} -m pip install flask
!pip install flask

In [1]:
from carsfactors import carsfactors
cf = carsfactors()

### Test Model first - Get stats

In [2]:
cf.model_stats()

random forest hyperparameters:
max_features: 0.5
n_estimators: 82
min_samples_leaf: 2
training/testing attributes:
trainint shape: (38521, 82)
test_size: 958
stratify train_test_split according to manufacturer-model combination: [145 145 122 ... 207 302  96]
training...
result...
score: 0.6163408223189843
RMSE -- train: 70.30512159806368, test: 83.32955417338682


'[0.6163408223189843, 70.30512159806368, 83.32955417338682]'

### Get Determination

infer for what we guess to be a fast sell

In [ ]:
cf.model_infer('BMW', 'automatic', 'black', 10000, 2019, 'gasoline', 5.7, 'suv',\
               True, 'all', 30000, 20)

infer for what we guess to be a slow sell

In [ ]:
cf.model_infer('Pontiac', 'mechanical', 'orange', 200000, 2000, 'diesel', 2.0, 'hatchback',\
               False, 'front', 90000, 0)

### Start up the service

In [ ]:
!python carfactors_service.py

Try out the links 
* [stats](http://localhost:8786/stats)
* [determination](http://localhost:8786/infer?manufacturer=Subaru&transmission=automatic&color=blue&odometer=12000&year=2020&engine_type=gasoline&engine_capacity=3.6&bodytype=suv&warranty=True&drivetrain=all&price=20000&numphotos=10)

URL formation for inference/determination:  
*http[]()://localhost:8786/infer?manufacturer=str&transmission=str&color=str&odometer=int&year=int&engine_type=int&engine_capacity=float&bodytype=str&warranty=bool&drivetrain=str&price=int&numphotos=int*

URL formation to see model statistics:  
*http[]()://localhost:8786/stats*

manufacturer options:  
['Subaru', 'LADA', 'Dodge', 'УАЗ', 'Kia', 'Opel', 'Москвич', 'Alfa Romeo', 'Acura', 'Dacia', 'Lexus', 'Mitsubishi', 'Lancia', 'Citroen', 'Mini', 'Jaguar', 'Porsche', 'SsangYong', 'Daewoo', 'Geely', 'ВАЗ', 'Fiat', 'Ford', 'Renault', 'Seat', 'Rover', 'Volkswagen', 'Lifan', 'Jeep', 'Cadillac', 'Audi', 'ЗАЗ', 'Toyota', 'ГАЗ', 'Volvo', 'Chevrolet', 'Great Wall', 'Buick', 'Pontiac', 'Lincoln', 'Hyundai', 'Nissan', 'Suzuki', 'BMW', 'Mazda', 'Land Rover', 'Iveco', 'Skoda', 'Saab', 'Infiniti', 'Chery', 'Honda', 'Mercedes-Benz', 'Peugeot', 'Chrysler']

transmission options:  
['automatic', 'mechanical']

color options:  
['silver', 'blue', 'red', 'black', 'grey', 'other', 'brown', 'white', 'green', 'violet', 'orange', 'yellow']

engine_type options:  
['gasoline', 'diesel', 'electric']

body_type options:  
['universal', 'suv', 'sedan', 'hatchback', 'liftback', 'minivan', 'minibus', 'van', 'pickup', 'coupe', 'cabriolet', 'limousine']

warranty options:  
True, False

drivetrain options:  
['all', 'front', 'rear']


### You must kill the kernel to try again for the port stays locked to the current kernel

# Summary
* Assignment and Model Results
* Techniques to improve the results

PUT YOUR ANSWERS HERE which could include additional cells with working code examples

The model has an R2 score of 0.616, which means %61.6 of the variance in the data is captured by the model.  This is considered good.  It is possible to achieve an R2 score as high as 0.85, but this results in an unacceptably overfitted model.

The model has an RMSE of 70.305 in training and 83.330 in testing.  These are both within the label standard deviation of 112.837.  Again we can achieve an RMSE as low as 42 in training, but this results in an unacceptably overfitted model.

Employing a fully connected neural network (FCN) could capture the nonlinear relationships between inputs and outputs.  This may require an architecture prohibitively large for CPU training (CPU training is necessarily the case using Sklearn).  If done in Tensorflow using a GPU, we could use 3 hidden layers with many neurons, and employ variance reduction in the form of dropout.